### DISCUSSION TITLE
Hanging tasks in the running state using DockerOpertator and StandardTaskRunner

### DISCUSSION THREAD

**Discussion by M3dv3D3V** [apache/airflow#40715] [Q&A]
Hello, on `Airflow 2.9.2` version, tasks are hanging in the running status using `DockerOpertator` and `StandardTaskRunner` as `task_runner`. There is a theory that freezes occur due to deadlocks, because one of the first messages in the logs talks about possible deadlocks if you use `os.fork()`, which is used in [StandardTaskRunner](https://github.com/apache/airflow/blob/main/airflow/task/task_runner/standard_task_runner.py#L71) (message below).
```python
{warnings.py:112} WARNING - /home/airflow/.local/lib/python3.12/site-packages/airflow/task/task_runner/standard_task_runner.py:61: DeprecationWarning: This process (pid=62100) is multi-threaded, use of fork() may lead to deadlocks in the child.
pid = os.fork()
```
The status of a hung task (process) - `S` (`sleeping`) in the `STAT` column, was obtained using the `ps -aux` command. An attempt to wake up the process using the `kill -CONT <PID>` command was unsuccessful.

After creating a new attempt at a hung task using the `Clear task` button in the Airflow UI, the previous task (process) goes into the `Z` - `zombie` state and remains in the system in the future. `Zombie` processes disappear only after a complete reboot of the system.

I will be glad to receive any comments and help in solving this problem. :pray:

**Comment by potiuk** [comment]
Speaking of zombie state, you might have similar problem as in https://github.com/apache/airflow/discussions/40596 where badly written Docker image (badly written init script) was not reaping the zombie processes correctly. You can read it and see if it causes your problems.

When it comes the threading warning, this is an interesting one, I've not seen that one before and we should take a look at that likely. I  don't recall that we start threads in Airlfow by default but it might be that there are other things at play. If you are using open-lineage. I believe some older versions of openlineage started some threads and that kind of warning **could** be caused by those threads starting before fork() happened. And I believe recently it was changed to multi-processing: @mobuchowski @kacpermuda -> WDYT? can you comment on that please?

**Reply by mobuchowski** [reply]
>And I believe recently it was changed to multi-processing: @mobuchowski @kacpermuda -> WDYT? can you comment on that please?

We're forking since very recently for task-level execution. We're not using threads in those versions though - we were using threads in the old `openlineage-airflow` package that does not work with Airflow 2.9. The fork the warning is talking about is earlier though - in the `StandardTaskRunner` in regular Airflow.

Would be nice to see if `OPENLINEAGE_DISABLE=true` would clear that - maybe there's some edge case here? We're not doing anything special for `DockerOperator`.

**Reply by M3dv3D3V** [reply]
We don't use `OPENLINEAGE_DISABLE=true`. Also, we don't use `openlineage` explicitly, but `apache-airflow-providers-openlineage` has version `1.8.0`.

**Reply by potiuk** [reply]
Just to rule it out - can you set it to `TRUE` it for a while and see if that solves the issue? Also - how the `init` investigation that I pointed you at is going ? Maybe the problem is that you have something even more "sophisticated" in your docker containert that is triggering it - in which case we will not ba able to help you before you look at all the specific things that you do (like probably bad `init` processs in your image) that you have to fix before there are any next steps for you to see if things are fixed.

**Reply by kacpermuda** [reply]
Just to clarify the typo, you can either set `AIRFLOW__OPENLINEAGE__DISABLED` to True or `OPENLINEAGE_DISABLED` (with a `D` at the end) to True.

**Reply by M3dv3D3V** [reply]
@potiuk, thank you for the investigation. I studied it. As far as I know, we do not run new scheduler processes inside `DockerOperator`. Also, we did not observe a large number of scheduler processes using `ps aux`.

**Reply by M3dv3D3V** [reply]
@kacpermuda, thank you. We will try to set `AIRFLOW__OPENLINEAGE__DISABLED` to `True` and look at the further state. I will also inform you about the result.

**Reply by potiuk** [reply]
I was talking about zombies that remain. This was your point above and indication that possibly (no idea about taht) Airflow ITSELF is run in the container with bad init. This is what should be investigated.

**Reply by potiuk** [reply]
Also - maybe the image you are using in your Docker container has wrong init or entrypoint (that would lead to zombies in the Docker container itself - you should always have a proper init process that propagates signals and reaps zombies. If you run your docker conteiner via docker operators without proper init, weird things happens nad resource usage might go through the roof.

**Reply by potiuk** [reply]
Generally - those zombies you mention indicate that the deployment of airflow or your docker operator is somewhat broken and it might cause ripple effect - so fixing it first should be the priority, otherwise you (not mentioning anyone who wants to help you) might loose a lot of time for chasing red herrings - we know that you have zombie processes remaining. And this means they are not properly reaped. Fix it first.

**Reply by M3dv3D3V** [reply]
Ok, we'll double-check our entrypoint. Perhaps you have the perfect entrypoint?

**Reply by potiuk** [reply]
See the links I posted in the discussion linked above. Airlfow provides a reference image with 'dumb-init" and the links there explain how signal propagation is done in it.

**Reply by potiuk** [reply]
https://github.com/apache/airflow/discussions/40596#discussioncomment-10009225 contains a lead description of the problem and links that you can follow to read more.

**Reply by M3dv3D3V** [reply]
@kacpermuda, just for reference. After installing `AIRFLOW__OPENLINEAGE__DISABLED=true`, tasks continue to hang.

**Reply by kacpermuda** [reply]
Ok, good to know, so as Jarek said, it might be some issues related to your airflow deployment or docker operator and looks like OpenLineage has nothing to do with it.

**Reply by M3dv3D3V** [reply]
We may be working on verifying our deployment right now. I would also like to point out that freezes don't always happen. There are times when tasks don't hang after a reboot.

**Reply by M3dv3D3V** [reply]
@potiuk, thanks for your advice with `dumb-init`. We slightly changed the call `/entrypoint` of the script ([as indicated in this discussion)](https://github.com/apache/airflow/discussions/25225) and the zombie processes disappeared. But unfortunately, the processes also continue to hang.

**Reply by M3dv3D3V** [reply]
I would also like to add that after a few hours the message below appears and the task continues to hang:
`{local_task_job_runner.py:163} INFO - Task is not able to be run`

**Reply by potiuk** [reply]
Impossible to help you. You need to  describe more details - including what and how you are running your Docker Operator and Standard Task Runner (but this is what everyone does) - It's likely you have some extra plugins or hooks or policies that start new threads before the fork happens - because this is not something that Airflow itself does.

So you should focus on either describing your case in detail or looking somewhere in the cusomisations, special deployment or anything that you customized in your Airflow deployment - because in no circumstances Airflow would start new threads before running a task. Ideally you can look at all your logs and see what is starting the threads - but that's something only you can do, possibly with tools like pyspy you can track what's going on and who calls what to find out.

This is what my guess is with the very limited information you provided.

**Comment by kacpermuda** [comment]
It would be good to know if you are indeed using OpenLineage. When it comes to OpenLineage, we had these two PRs: [Snowflake](https://github.com/apache/airflow/pull/39113) and [Redshift](https://github.com/apache/airflow/pull/40197) that may have been causing deadlocks. We later switched to `ProcessPoolExecutor` in this [PR](https://github.com/apache/airflow/pull/39235) for DAG level events and the latest [PR](https://github.com/apache/airflow/pull/40078) executes whole OL extraction in separate processes.

EDIT: I did not notice you mentioned DockerOperator, so not all i wrote makes sense.
